rust/blockstore/src/arrow/block/delta/storage.rs

```diff
@@ -185,31 +185,89 @@ impl BlockStorage {
         }
     }
 
-    pub fn split<K: ArrowWriteableKey>(&self, split_size: usize) -> (CompositeKey, BlockStorage) {
+    pub fn split_before_size<K: ArrowWriteableKey>(
```

Comment:
I don't like the separation of these two methods. I did this because the ordered and unordered modes expect different split results.

For the unordered writer mode, it always splits deltas in half. This means that the split should include one key over the limit; otherwise the effective size limit for K/V pairs is half the actual maximum. Splitting the delta at the maximum size instead can lead to excessive fragmentation with our current writing logic. E.g.:

- 1 full block with key range A-B
- 1 partial block with key range Z-Z
- Inserts for keys C-Y go to the full block (A-B), then are split off into a new block (each key becomes a new block if the keys are inserted in roughly Y-C order)

I am open to changing the behavior of the unordered writer to fix the fragmentation issue, just wanted to bring it up first.

Reply:
discussed with @rescrv that this is ok to leave as-is for now and improving fragmentation/write amplification requires more thought behind it